llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp

```diff
@@ -2495,13 +2495,17 @@ Value *LibCallSimplifier::optimizeLog(CallInst *Log, IRBuilderBase &B) {
 
   // log(pow(x,y)) -> y*log(x)
   AttributeList NoAttrs;
-  if (ArgLb == PowLb || ArgID == Intrinsic::pow) {
+  if (ArgLb == PowLb || ArgID == Intrinsic::pow || ArgID == Intrinsic::powi) {
     Value *LogX =
         Log->doesNotAccessMemory()
             ? B.CreateCall(Intrinsic::getDeclaration(Mod, LogID, Ty),
                            Arg->getOperand(0), "log")
             : emitUnaryFloatFnCall(Arg->getOperand(0), TLI, LogNm, B, NoAttrs);
-    Value *MulY = B.CreateFMul(Arg->getArgOperand(1), LogX, "mul");
+    Value *Y = Arg->getArgOperand(1);
+    // If power is integer constant, convert to FP for FMul.
+    if (ConstantInt *C = dyn_cast<ConstantInt>(Y))
+      Y = ConstantFP::get(Ty, C->getSExtValue());
```

Comment:
If the `exponent` is not a constant, you will create an illegal fmul with an integer LHS.
Please add the following negative test and pre-commit these tests in a separate commit.
```
define double @log_pow_nonconst(double %x, i32 %y) {
  %pow = call fast double @llvm.powi.f64.i32(double %x, i32 %y)
  %log = call fast double @log(double %pow)
  ret double %log
}
```

Another question: why don't we fold `log(powi(x,y)) -> (sitofp y) * log(x)` even if `Y` is not a constant? I think it is also profitable if `powi` is only used by the `log`.

Reply:
@dtcxzyw I did a standard sitofp initially, but that adds a casting instruction right? In the compiled asm there's an addition `scvtf` if using a non-constant cast.

I'll fix the illegal fmul, sorry about that.

Reply:
> @dtcxzyw I did a standard sitofp initially, but that adds a casting instruction right? In the compiled asm there's an addition `scvtf` if using a non-constant cast.

GCC doesn't apply the transform if `y` is not a constant: https://godbolt.org/z/qTd6Tnh3c
If the `powi` is only used by the `log`, `sitofp + fmul` is definitely cheaper than a call to `pow`.

Reply:
Got it! Will update PR  with the changes.

Reply:
Added sitofp conversion for non-consts.